lib/compat/wordpress-6.5/class-gutenberg-rest-global-styles-revisions-controller-6-5.php

```diff
@@ -0,0 +1,102 @@
+<?php
+/**
+ * REST API: Gutenberg_REST_Global_Styles_Revisions_Controller class, inspired by WP_REST_Revisions_Controller.
+ *
+ * @package WordPress
+ * @subpackage REST_API
+ * @since 6.3.0
+ */
+
+/**
+ * Core class used to access global styles revisions via the REST API.
+ *
+ * @since 6.3.0
+ *
+ * @see WP_REST_Controller
+ */
+class Gutenberg_REST_Global_Styles_Revisions_Controller_6_5 extends Gutenberg_REST_Global_Styles_Revisions_Controller_6_4 {
+	/**
+	 * Registers the controller's routes.
+	 *
+	 * @since 6.3.0
+	 * @since 6.5.0 Adds route to fetch individual global styles revisions.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->parent_base . '/(?P<parent>[\d]+)/' . $this->rest_base . '/(?P<id>[\d]+)',
+			array(
+				'args'   => array(
+					'parent' => array(
+						'description' => __( 'The ID for the parent of the global styles revision.' ),
+						'type'        => 'integer',
+					),
+					'id'     => array(
+						'description' => __( 'Unique identifier for the global styles revision.' ),
+						'type'        => 'integer',
+					),
+				),
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_item' ),
+					'permission_callback' => array( $this, 'get_item_permissions_check' ),
+					'args'                => array(
+						'context' => $this->get_context_param( array( 'default' => 'view' ) ),
+					),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+		parent::register_routes();
+	}
+
+	/**
+	 * Retrieves one global styles revision from the collection.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_item( $request ) {
+		$parent = $this->get_parent( $request['parent'] );
+		if ( is_wp_error( $parent ) ) {
+			return $parent;
+		}
+
+		$revision = $this->get_revision( $request['id'] );
```

Comment:
In the `get_items` version of retrieving revisions, they're guarded behind a check against the `$parent` of `if ( wp_revisions_enabled( $paren ) ) {`... should we add a similar check here, too?

Reply:
I've modelled this class as closely as possible to [WP_REST_Revisions_Controller](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php).

The reason I wouldn't want to make these changes here is because they should be probably be raised in trac and discussed in the wider context of revisions.

Even so I think it should be okay. 

This is what the docs say about the `supports` flag for [register_post_type](https://developer.wordpress.org/reference/functions/register_post_type/), specifically: 

> 'revisions' feature dictates whether the post type will store revisions

So if a post type cannot store revisions, there'll be none to fetch.

It's possible `wp_revisions_enabled` is used in `get_items` to save an expensive `new WP_Query()` call.

Reply:
> I've modelled this class as closely as possible to WP_REST_Revisions_Controller.

@ramonjd I'm curious why you couldn't just **use** the core revisions controller, what feature was missing?

Reply:
@adamsilverstein Thanks for the question. It is a good one. 

Global styles post revisions store stringified JSON in the post content field. When preparing the response, the JSON needs to be parsed and structured according to the theme.json model, that is, something like this:

```json

{
     "settings": {},
     "styles: {},
     ... // whatever else comes along
}
```

That's the main difference - the responses are different.

I did consider extending `WP_REST_Revisions_Controller` way back when I introduced the [global styles revisions endpoint](https://github.com/WordPress/gutenberg/pull/49974#discussion_r1174845482).

Global styles revisions were a bit new and had the following characteristics:

- Didn't, and still doesn't, return excerpt, content, guid fields (has 2 custom fields: settings, styles and later ??)
- Did not (yet) require delete routes
- Had to do some JSON parsing when preparing the response

Back then, it was more code to override the parent class's functions, and I thought having it as a stand alone class left more elbow room for subsequent tweaks and changes. Especially since global styles revisions were new and evolving.

Also, since I tried to sync what I needed from `WP_REST_Revisions_Controller`, I thought that it wouldn't affect backwards compatibility if we one day decided to refactor this code and extend `WP_REST_Revisions_Controller`. 

Today I think the functionality is more balanced. Overriding fields and response-related methods would still be needed, but are a lot of shared methods between the two classes and, arguably, a delete route would bring it in line with current revision expectations.

What do you think?

Reply:
Hey @ramonjd thanks for clarifying your thinking here. I agree it makes sense to build out a separate class for your exact needs, especially while building something new where you want to change things and avoid breaking the existing controller.

> That's the main difference - the responses are different.

Given that you mainly want to alter the response behavior, I wonder if you could use the existing revisions controller and leverage the `rest_prepare_revision` filter provided to adjust the shape of the return as you desire - this should let you control the exact response you want for each revision.

My main concern with adding another controller is the added maintenance burden. For example the issue you discovered with mismatched parent ids would need to get fixed in two places.

If you feel your controller is doing everything it needs to, I would encourage at least experimenting with using the filter approach and relying on the existing controller. This should reduce the amount of code you need.  

cc: @TimothyBJacobs for a second opinion

Reply:
Thanks, I appreciate the insight. 🙇🏻 

> If you feel your controller is doing everything it needs to, I would encourage at least experimenting with using the filter approach and relying on the existing controller.

Good call. I'd like to try this. The maintenance burden is on my mind 😅

Reply:
+1 let me know if you find any shortcomings in the core endpoint or the filter/hooks it provides -  if so maybe we can address those in core.

Reply:
I have a Core PR in progress here:

- https://github.com/WordPress/wordpress-develop/pull/5699

I ended up replacing the base class with `WP_REST_Revisions_Controller` so that the `WP_REST_Global_Styles_Revisions_Controller` was still available (I wasn't sure about how radical I should be about maintaining backwards compatibility), but more importantly, so I could preserve the following features by overwriting class methods:

- no deletion of revisions (I think this can come later - probably needs a discussion with Gutenberg Core members)
- massage the schema

Thanks for the nudge! It removes a lot of code.